ext/user_request.c

```diff
@@ -0,0 +1,168 @@
+#include "user_request_arginfo.h"
+#include "zend_API.h"
+
+#define NS "DDTrace\\UserRequest\\"
+
+static struct {
+    ddtrace_user_req_listeners **listeners;
+    size_t size;
+} reg_listeners;
+
+DDTRACE_PUBLIC bool ddtrace_user_req_add_listeners(ddtrace_user_req_listeners *listeners)
+{
+    if (strcmp(sapi_module.name, "cli") != 0) {
+        return false;
+    }
+
+    reg_listeners.size += 1;
+    reg_listeners.listeners = realloc(reg_listeners.listeners,
+        sizeof(*reg_listeners.listeners) * reg_listeners.size);
+    reg_listeners.listeners[reg_listeners.size - 1] = listeners;
+
+    return true;
+}
+
+PHP_FUNCTION(DDTrace_UserRequest_has_listeners)
+{
+    if (zend_parse_parameters_none() == FAILURE) {
+        RETURN_FALSE;
+    }
+
+    RETURN_BOOL(reg_listeners.size > 0);
+}
+
+PHP_FUNCTION(DDTrace_UserRequest_notify_start)
+{
+    zend_object *span;
+    zend_array *array;
+
+    ZEND_PARSE_PARAMETERS_START(2, 2)
+        Z_PARAM_OBJ_OF_CLASS_EX(span, ddtrace_ce_root_span_data, 0, 1)
+        Z_PARAM_ARRAY_HT(array)
+    ZEND_PARSE_PARAMETERS_END();
+
+    ddtrace_span_data *span_data = OBJ_SPANDATA(span);
+    if (span_data->duration != 0) {
+        php_error_docref(NULL, E_WARNING, "Span already finished");
+        RETURN_NULL();
+    }
+    if (span_data->notify_user_req_end) {
+        php_error_docref(NULL, E_WARNING, "Start of span already notified");
+        RETURN_NULL();
+    }
+
+    zend_array *replacement_resp = NULL;
+    for (size_t i = 0; i < reg_listeners.size; i++) {
+        ddtrace_user_req_listeners *listener = reg_listeners.listeners[i];
+        zend_array *repl = listener->start_user_req(listener, span, array);
+        if (repl != NULL && replacement_resp == NULL) {
+            replacement_resp = repl;
+        } else if (repl != NULL) {
+#ifdef GC_TRY_DELREF
+            GC_TRY_DELREF(repl);
+#else
+            GC_DELREF(repl);
+#endif
+            if (GC_REFCOUNT(repl) == 0) {
+                zend_array_destroy(repl);
+            }
+        }
+    }
+
+    span_data->notify_user_req_end = true;
+
+    if (replacement_resp != NULL) {
+        RETURN_ARR(replacement_resp);
+    } else {
+        RETURN_NULL();
+    }
+}
+
+PHP_FUNCTION(DDTrace_UserRequest_notify_commit)
+{
+    zend_object *span;
+    zend_long status;
+    zend_array *headers;
+
+    ZEND_PARSE_PARAMETERS_START(3, 3)
+        Z_PARAM_OBJ(span)
+        Z_PARAM_LONG(status)
+        Z_PARAM_ARRAY_HT(headers)
+    ZEND_PARSE_PARAMETERS_END();
+
+    zend_array *replacement_resp = NULL;
+    for (size_t i = 0; i < reg_listeners.size; i++) {
+        ddtrace_user_req_listeners *listener = reg_listeners.listeners[i];
+        zend_array *repl = listener->response_committed(listener, span, status, headers);
+        if (repl != NULL && replacement_resp == NULL) {
+            replacement_resp = repl;
+        } else if (repl != NULL) {
+#ifdef GC_TRY_DELREF
+            GC_TRY_DELREF(repl);
+#else
+            GC_DELREF(repl);
```

Comment:
This doesn't sound safe to me? I think you should instead polyfill GC_TRY_DELREF() in compatibility.h and check for `GC_FLAGS() & IS_ARRAY_IMMUTABLE`.

Reply:
Anyway, a better suggestion would be zend_array_release(), which does the delref and rc check for you. And we already polyfill that one.